File: Chapter 5/5.3-spinnycube/ConfigureEnvironment/GLCoreProfileView.m
//
//  GLCoreProfileView.m
//  ConfigureEnvironment
//
//  Created by 陈杰 on 26/10/2017.
//  Copyright © 2017 陈杰. All rights reserved.
//


#import "GLCoreProfileView.h"
//#import <OpenGL/gl3.h>
#import <OpenGL/OpenGL.h>
#import <GLKit/GLKit.h>

@interface GLCoreProfileView()
@property (atomic, strong) NSTimer *lifeTimer;
@property (atomic, assign) CGFloat lifeDuration;

@property (atomic, assign) GLuint program;
@property (atomic, assign) GLuint vertexArray;
@property (atomic, assign) GLuint vertexArrayBuffer;

@property (atomic, assign) GLint mv_location;
@property (atomic, assign) GLint proj_location;
@property (atomic, assign) GLKMatrix4 mv_Matrix;
@property (atomic, assign) GLKMatrix4 proj_Matrix;
@end

@implementation GLCoreProfileView
#pragma mark - lifecycle methods
- (instancetype)initWithCoder:(NSCoder *)decoder {
    NSOpenGLPixelFormatAttribute pixelFormatAttributes[] = {
        NSOpenGLPFAColorSize, 32,
        NSOpenGLPFADepthSize, 24,
        NSOpenGLPFAStencilSize, 8,
        NSOpenGLPFAAccelerated,
        NSOpenGLPFAOpenGLProfile, NSOpenGLProfileVersion4_1Core,
        0
    };
    NSOpenGLPixelFormat *pixelFormat = [[NSOpenGLPixelFormat alloc] initWithAttributes:pixelFormatAttributes];
    NSOpenGLContext *openGLContext = [[NSOpenGLContext alloc] initWithFormat:pixelFormat shareContext:nil];
    if (self = [super initWithCoder:decoder]) {
        [self setOpenGLContext:openGLContext];
        [self.openGLContext makeCurrentContext];
        
        _lifeTimer = [NSTimer scheduledTimerWithTimeInterval:0.02 target:self selector:@selector(lifeTimerUpdate) userInfo:nil repeats:YES];
    }
    return self;
}

- (void)dealloc {
    [_lifeTimer invalidate];
    _lifeTimer = nil;
    
    glDeleteVertexArrays(1, &_vertexArray);
    glDeleteProgram(_program);
    glDeleteBuffers(1, &_vertexArrayBuffer);
}

- (void)prepareOpenGL {
    NSLog(@"Version: %s", glGetString(GL_VERSION));
    NSLog(@"Renderer: %s", glGetString(GL_RENDERER));
    NSLog(@"Vendor: %s", glGetString(GL_VENDOR));
    NSLog(@"GLSL Version: %s", glGetString(GL_SHADING_LANGUAGE_VERSION));
    
    [self loadShaders];
    
    static const GLfloat vertex_positions[] = {
        -0.25f,  0.25f, -0.25f,      -0.25f, -0.25f, -0.25f,     0.25f, -0.25f, -0.25f,
        0.25f, -0.25f, -0.25f,       0.25f,  0.25f, -0.25f,      -0.25f,  0.25f, -0.25f,

        -0.25f, -0.25f, 0.25f,       0.25f, -0.25f, 0.25f,       0.25f, -0.25f, -0.25f,
        0.25f, -0.25f, -0.25f,       -0.25f, -0.25f, -0.25f,     -0.25f, -0.25f, 0.25f,

        0.25f, 0.25f, 0.25f,         0.25f, 0.25f, -0.25f,       0.25f, -0.25f, -0.25f,
        0.25f, -0.25f, -0.25f,       0.25f, -0.25, 0.25f,        0.25f, 0.25f, 0.25f,

        -0.25f, -0.25f, 0.25f,       -0.25f, 0.25f, 0.25f,       -0.25f, -0.25f, -0.25f,
        -0.25f, 0.25f, -0.25f,       -0.25f, -0.25f, -0.25f,     -0.25f, 0.25f, 0.25f,

        0.25f, 0.25f, 0.25f,         0.25f, -0.25f, 0.25f,       -0.25f, -0.25f, 0.25f,
        -0.25f, -0.25f, 0.25f,       -0.25f, 0.25f, 0.25f,       0.25f, 0.25f, 0.25f,

        -0.25f,  0.25f, -0.25f,       0.25f,  0.25f, -0.25f,      0.25f,  0.25f,  0.25f,
        0.25f,  0.25f,  0.25f,        -0.25f,  0.25f,  0.25f,     -0.25f,  0.25f, -0.25f
    };
    
    // Now generate some data and put it in a buffer object
    glGenBuffers(1, &_vertexArrayBuffer);
    glBindBuffer(GL_ARRAY_BUFFER, _vertexArrayBuffer);
    glBufferData(GL_ARRAY_BUFFER, sizeof(vertex_positions), vertex_positions, GL_STATIC_DRAW);
    
    glGenVertexArrays(1, &_vertexArray);
    glBindVertexArray(_vertexArray);
    
    // Set up our vertex attribute
    glEnableVertexAttribArray(GLKVertexAttribPosition);
    glVertexAttribPointer(GLKVertexAttribPosition, 3, GL_FLOAT, GL_FALSE, 0, NULL);
    
    glEnable(GL_DEPTH_TEST);
}

- (void)reshape {
    NSRect bounds = [self bounds];
    glViewport(0, 0, NSWidth(bounds), NSHeight(bounds));
    GLfloat aspect = NSWidth(bounds)/NSHeight(bounds);
    _proj_Matrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(65.0), aspect, 0.1f, 1000.0f);
}

- (void)drawRect:(NSRect)dirtyRect {
    static const GLfloat green[] = { 0.0f, 0.15f,
        0.0f, 1.0f };
    static const GLfloat depth = 1.0f;
    glClearBufferfv(GL_COLOR, 0, green);
    glClearBufferfv(GL_DEPTH, 0, &depth);
    
    glUseProgram(_program);
    glUniformMatrix4fv(_proj_location, 1, GL_FALSE, _proj_Matrix.m);
    for (int i = 0; i < 24; i++) {
        GLfloat f = _lifeDuration*0.3f + i;
        GLKMatrix4 xRotateMatrix = GLKMatrix4MakeRotation(_lifeDuration*GLKMathDegreesToRadians(21.0f), 1.0f, 0.0f, 0.0f);
        GLKMatrix4 yRotateMatrix = GLKMatrix4MakeRotation(_lifeDuration*M_PI_4, 0.0f, 1.0f, 0.0f);
        GLKMatrix4 translateMatrix = GLKMatrix4MakeTranslation(sinf(2.1f*f)*2.0f, cosf(1.7f*f)*2.0f, sinf(1.3f*f)*cosf(1.5f*f)*2.0f);
        GLKMatrix4 modelWroldMatrix = GLKMatrix4Multiply(GLKMatrix4Multiply(translateMatrix, yRotateMatrix), xRotateMatrix);
        GLKMatrix4 viewMatrix = GLKMatrix4MakeTranslation(0.0f, 0.0f, -7.0f);
        _mv_Matrix = GLKMatrix4Multiply(viewMatrix, modelWroldMatrix);
        glUniformMatrix4fv(_mv_location, 1, GL_FALSE, _mv_Matrix.m);
        glDrawArrays(GL_TRIANGLES, 0, 36);
    }
    glFlush();
}

#pragma mark - public methods

#pragma mark - private methods
- (BOOL)loadShaders {
    GLuint vertexShader;
    GLuint fragShader;
    NSString *vertexShaderPathName;
    NSString *fragShaderPathName;
    
    _program = glCreateProgram();
    
    vertexShaderPathName = [[NSBundle mainBundle] pathForResource:@"Shader" ofType:@"vsh"];
    if (![self compileShader:&vertexShader type:GL_VERTEX_SHADER filePath:vertexShaderPathName]) {
        NSLog(@"Failed to compile vertex shader");
    }
    
    fragShaderPathName = [[NSBundle mainBundle] pathForResource:@"Shader" ofType:@"fsh"];
    if (![self compileShader:&fragShader type:GL_FRAGMENT_SHADER filePath:fragShaderPathName]) {
        NSLog(@"Failed to compile fragment shader");
    }
    
    glAttachShader(_program, vertexShader);
    glAttachShader(_program, fragShader);

    glBindAttribLocation(_program, GLKVertexAttribPosition, "position");
    
    if (vertexShader != 0) {
        glDeleteShader(vertexShader);
        vertexShader = 0;
    }
    if (fragShader != 0) {
        glDeleteShader(fragShader);
        fragShader = 0;
    }
    
    if (![self linkProgram:_program]) {
        NSLog(@"Failed to link program: %d", _program);
        if (_program != 0) {
            glDeleteProgram(_program);
            _program = 0;
        }
        return NO;
    }

    _mv_location = glGetUniformLocation(_program, "mv_matrix");
    _proj_location = glGetUniformLocation(_program, "proj_matrix");
    return YES;
}

- (BOOL)compileShader:(GLuint *)shader type:(GLenum)type filePath:(NSString *)path {
    const GLchar *shaderSource = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil].UTF8String;
    *shader = glCreateShader(type);
    glShaderSource(*shader, 1, &shaderSource, nil);
    glCompileShader(*shader);
    
    GLint status = 0;
    glGetShaderiv(*shader, GL_COMPILE_STATUS, &status);
    if (status == 0) {
        GLint logLen = 0;
        glGetShaderiv(*shader, GL_INFO_LOG_LENGTH, &logLen);
        GLchar *infoLog = malloc(sizeof(char) * logLen);
        glGetShaderInfoLog(*shader, logLen, NULL, infoLog);
        NSLog(@"Shader at: %@", path);
        fprintf(stderr, "Info Log: %s\n", infoLog);
        
        glDeleteShader(*shader);
        return NO;
    }
    return YES;
}

- (BOOL)linkProgram:(GLuint)program {
    glLinkProgram(program);
    GLint status = 0;
    glGetProgramiv(program, GL_LINK_STATUS, &status);
    if (status == 0) {
        GLint logLen = 0;
        glGetProgramiv(program, GL_INFO_LOG_LENGTH, &logLen);
        GLchar *infoLog = malloc(sizeof(char) * logLen);
        glGetProgramInfoLog(program, logLen, NULL, infoLog);
        fprintf(stderr, "Prog Info Log: %s\n", infoLog);
        return NO;
    }
    return YES;
}

#pragma mark - listening methods
- (void)lifeTimerUpdate {
    _lifeDuration += _lifeTimer.timeInterval;
    // _lifeDuration为程序运行时间
    [self drawRect:self.bounds];
}

#pragma mark - accessor methods

@end

File: 1_Introduction_MacOSEvironment/4_FragmentShader_start/FragmentShader/GLCoreProfileView.swift
//
//  GLCoreProfileView.swift
//  FragmentShader
//
//  Created by chenjie on 2019/4/27.
//  Copyright © 2019 starrythrone. All rights reserved.
//

import Cocoa
import OpenGL.GL3

class GLCoreProfileView: NSOpenGLView {

    //MARK: - Propeties
    fileprivate var lifeDuration: CGFloat = 0
    fileprivate var program: GLuint = 0
    fileprivate var vertexArray: GLuint = 0
    
    //MARK: - Life Cycle
    required init?(coder decoder: NSCoder) {
        super.init(coder: decoder)
        
        let openGLContextConfigured = setupOpenGLContext()
        if !openGLContextConfigured {
            print("Prepare OpenGLContextFailed...")
        }
    }
    
    deinit {
        glDeleteProgram(program)
        glDeleteVertexArrays(1, &vertexArray)
    }

    //MARK: - Override
    override func prepareOpenGL() {
        super.prepareOpenGL()
        
        if let versionString = glGetString(GLenum(GL_VERSION)) {
            print("GL_VERSION: " + String(cString: versionString))
        }
        if let renderString = glGetString(GLenum(GL_RENDERER)) {
            print("GL_RENDERER: " + String(cString: renderString))
        }
        if let vendorString = glGetString(GLenum(GL_VENDOR)) {
            print("GL_VENDOR: " + String(cString: vendorString))
        }
        if let glVersionString = glGetString(GLenum(GL_SHADING_LANGUAGE_VERSION)) {
            print("GL_SHADING_LANGUAGE_VERSION: " + String(cString: glVersionString))
        }
        
        let shaderLoaded = loadShaders()
        if !shaderLoaded {
            return
        }
        
        glGenVertexArrays(1, &vertexArray);
        glBindVertexArray(vertexArray);
    }
    
    override func reshape() {
        super.reshape()
        let bounds = self.bounds
        glViewport(0, 0, GLsizei(NSWidth(bounds)), GLsizei(NSHeight(bounds)));
    }
    
    
    override func draw(_ dirtyRect: NSRect) {
        super.draw(dirtyRect)
    }
    
    //MARK: - Private Methods
    fileprivate func setupOpenGLContext() -> Bool {
        var pixelFormatAttributes: [NSOpenGLPixelFormatAttribute] =
            [UInt32(NSOpenGLPFAColorSize), 32,
             UInt32(NSOpenGLPFAAccelerated),
             UInt32(NSOpenGLPFAOpenGLProfile), UInt32(NSOpenGLProfileVersion4_1Core), 0]
        guard let pixelFormat = NSOpenGLPixelFormat(attributes: &pixelFormatAttributes) else {
            return false
        }
        
        let currentOpenGLContext = NSOpenGLContext(format: pixelFormat, share: nil)
        self.openGLContext = currentOpenGLContext
        self.openGLContext?.makeCurrentContext()
        return true
    }
    
    fileprivate func loadShaders() -> Bool {
        self.program = glCreateProgram()
        
        var vertexShader: GLuint = 0
        guard let vertexShaderPath = Bundle.main.path(forResource: "ShaderV", ofType: "vsh") else {
            print("Can not load vertexShader file")
            return false
        }
        let vertexShaderCompiled = compileShader(shader: &vertexShader, type: GLenum(GL_VERTEX_SHADER), filePath: vertexShaderPath)
        if (!vertexShaderCompiled) {
            return false
        }
        
        var fragmentShader: GLuint = 0
        guard let fragmentShaderPath = Bundle.main.path(forResource: "ShaderF", ofType: "vsh") else {
            print("Can not load fragmentShader file")
            return false
        }
        let fragmentShaderCompiled = compileShader(shader: &fragmentShader, type: GLenum(GL_FRAGMENT_SHADER), filePath: fragmentShaderPath)
        if (!fragmentShaderCompiled) {
            return false
        }
        
        glAttachShader(self.program, vertexShader)
        glAttachShader(self.program, fragmentShader)
        
        if (vertexShader != 0) {
            glDeleteShader(vertexShader);
            vertexShader = 0
        }
        if (fragmentShader != 0) {
            glDeleteShader(fragmentShader);
            fragmentShader = 0
        }
        
        let programLinked = linkProgram(program)
        if programLinked {
            return true
        } else {
            return false
        }
    }
    
    fileprivate func compileShader(shader: inout GLuint, type: GLenum, filePath: String) -> Bool {
        var source: UnsafePointer<Int8>?
        do {
            source = try NSString(contentsOfFile: filePath, encoding: String.Encoding.utf8.rawValue).utf8String
        } catch {
            print("Failed to load shader string...")
            return false
        }
        var shaderSource = UnsafePointer<GLchar>(source)
        if (shaderSource == nil) {
            return false
        }
        
        shader = glCreateShader(type)
        glShaderSource(shader, 1, &shaderSource, nil)
        glCompileShader(shader)
        
        var status: GLint = 0
        glGetShaderiv(shader, GLenum(GL_COMPILE_STATUS), &status)
        if status == 0 {
            var logLength: GLint = 0
            glGetShaderiv(shader, GLenum(GL_INFO_LOG_LENGTH), &logLength)
            var infoLog = [GLchar](repeating: 0, count: Int(logLength))
            glGetShaderInfoLog(shader, logLength, nil, &infoLog)
            let infoLogString = String(NSString.init(utf8String: &infoLog) ?? "")
            print("Compile shader failed at file:\n" + filePath + "\nwith log\n" + infoLogString)
            glDeleteShader(shader)
            return false
        }
        return true
    }
    
    fileprivate func linkProgram(_ program: GLuint) -> Bool {
        glLinkProgram(program)
        var status: GLint = 0
        glGetProgramiv(program, GLenum(GL_LINK_STATUS), &status)
        if status == 0 {
            var infoLength: GLint = 0
            glGetProgramiv(program, GLenum(GL_INFO_LOG_LENGTH), &infoLength)
            var infoLog = [GLchar](repeating: 0, count: Int(infoLength))
            glGetProgramInfoLog(program, infoLength, nil, &infoLog)
            let infoLogString = String(NSString.init(utf8String: &infoLog) ?? "")
            print("Like program fiaied with log: " + infoLogString)
            return false
        }
        return true
    }
}
